cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1545,6 +1559,13 @@ class ObjectInputFile final : public io::RandomAccessFile {
 // (for rational, see: https://github.com/apache/arrow/issues/34363)
 static constexpr int64_t kPartUploadSize = 10 * 1024 * 1024;
 
+// Above this threshold, use a multi-part upload instead of a single request upload.
+static constexpr int64_t kMultiPartUploadThresholdSize = 5 * 1024 * 1024;
```

Comment:
Should we config hard limit like this, or Switch to multipart when there're multi parts? (size > 10MiB during append)

Reply:
Would also be a valid approach and would simplify the code a bit. AWS S3 docs themselves state that you could/should use `PutObject` request for data up to 100MB (see [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)), so orienting at the part size which is 10MB would also be fine. What do other think?

Reply:
Both is ok for me, here S3OutputStream might manage these part of memory in memory buffer, so I think it shouldn't be too large? changing it default 10MB or making it tunable later is both ok for me

Reply:
[78efe6e](https://github.com/apache/arrow/pull/41564/commits/78efe6e48e11c198073a5912714c9be94421b283)